temporal-sdk/src/main/java/io/temporal/client/WorkflowFailedException.java

```diff
@@ -50,23 +59,72 @@ public RetryState getRetryState() {
     return retryState;
   }
 
+  /**
+   * This value is defined only if workflow failure is caused by an explicit WORKFLOW_TASK_COMPLETED
+   * command. If workflow timed out, was cancelled or terminated, this value is undefined.
+   *
+   * @return eventId of the WORKFLOW_TASK_COMPLETED event that reported (caused)
+   *     WORKFLOW_EXECUTION_FAILED command. -1 if undefined.
+   */
   public long getWorkflowTaskCompletedEventId() {
     return workflowTaskCompletedEventId;
   }
 
-  public static String getMessage(
+  /**
+   * Returns Event Type that caused {@code this} exception. This Event Type should be one of
+   *
+   * <ul>
+   *   <li>{@link EventType#EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED}
+   *   <li>{@link EventType#EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED}
+   *   <li>{@link EventType#EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT}
+   *   <li>{@link EventType#EVENT_TYPE_WORKFLOW_EXECUTION_FAILED}
+   * </ul>
+   *
+   * @return event type that caused {@code this} exception
+   */
+  public EventType getWorkflowCloseEventType() {
+    return workflowCloseEventType;
+  }
+
+  private static String getMessage(
       WorkflowExecution workflowExecution,
       String workflowType,
+      EventType closeEventType,
       long workflowTaskCompletedEventId,
       RetryState retryState) {
-    return "workflowId='"
+    return "Workflow execution "
```

Comment:
This would look much nicer with Guava's `MoreObjects.ToStringHelper` :-)

Reply:
We try our best to don't use Guava where it's not absolutely needed :) Also, this one is not a standard toString, it's a message and it's customized a little.

Reply:
I find even a string builder is easier to read than all these concatenations (granted that's what the JVM compiles this down to anyways)